cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -1906,23 +1913,54 @@ pub(crate) fn emit(
             sink.bind_label(done);
         }
 
-        Inst::XmmRmRImm { op, src, dst, imm } => {
-            let prefix = match op {
-                SseOpcode::Cmpps => LegacyPrefixes::None,
-                SseOpcode::Cmppd => LegacyPrefixes::_66,
-                SseOpcode::Cmpss => LegacyPrefixes::_F3,
-                SseOpcode::Cmpsd => LegacyPrefixes::_F2,
+        Inst::XmmRmRImm {
+            op,
+            src,
+            dst,
+            imm,
+            w,
+        } => {
+            let (prefix, opcode, len) = match op {
+                SseOpcode::Cmpps => (LegacyPrefixes::None, 0x0FC2, 2),
+                SseOpcode::Cmppd => (LegacyPrefixes::_66, 0x0FC2, 2),
+                SseOpcode::Cmpss => (LegacyPrefixes::_F3, 0x0FC2, 2),
+                SseOpcode::Cmpsd => (LegacyPrefixes::_F2, 0x0FC2, 2),
+                SseOpcode::Insertps => (LegacyPrefixes::_66, 0x0F3A21, 3),
+                SseOpcode::Pinsrb => (LegacyPrefixes::_66, 0x0F3A20, 3),
+                SseOpcode::Pinsrw => (LegacyPrefixes::_66, 0x0FC4, 2),
+                SseOpcode::Pinsrd => (LegacyPrefixes::_66, 0x0F3A22, 3),
+                SseOpcode::Pextrb => (LegacyPrefixes::_66, 0x0F3A14, 3),
+                SseOpcode::Pextrw => (LegacyPrefixes::_66, 0x0FC5, 2),
+                SseOpcode::Pextrd => (LegacyPrefixes::_66, 0x0F3A16, 3),
+                SseOpcode::Pshufd => (LegacyPrefixes::_66, 0x0F70, 2),
                 _ => unimplemented!("Opcode {:?} not implemented", op),
             };
-            let opcode = 0x0FC2;
-            let rex = RexFlags::clear_w();
+            let rex = if *w {
+                RexFlags::set_w()
+            } else {
+                RexFlags::clear_w()
+            };
             match src {
                 RegMem::Reg { reg } => {
-                    emit_std_reg_reg(sink, prefix, opcode, 2, dst.to_reg(), *reg, rex);
+                    if *op == SseOpcode::Pextrb || *op == SseOpcode::Pextrd {
```

Comment:
Little style preference mostly, but maybe do a `let swapped = match *op { ... };` and then `let (dst, src) = if swapped { ... } else { ... };`? This keeps the code duplication down in case we add more logic around the `emit_...()` call later.